### DISCUSSION TITLE
How to add custom theme variants to ThemeDictionaries?

### DISCUSSION THREAD

**Discussion by mysticpillow** [AvaloniaUI/Avalonia#11953] [Q&A] [answered]
How to add multiple custom themeVariants exported from https://theme.xaml.live/ to Application ThemeDictionaries?
How to switch between these theme variants in runtime by settings Application.Current.RequestedThemeVariant?
I simply cannot find any examples how to do this
What I've tried:
- Created a static class `CommonThemeVariants` with theme variants: `LightBlue` inherited from `Light`, `DarkBlue` inherited from `Dark`
- Set application resources like this:
```
  <Application.Resources>
    <ResourceDictionary>
      <ResourceDictionary.ThemeDictionaries>
        <ResourceDictionary x:Key="Default">
            <ColorPaletteResources x:Key="Default" ... />
        </ResourceDictionary>
        <ResourceDictionary x:Key="Dark">
            <ColorPaletteResources x:Key="Dark" ... />
        </ResourceDictionary>
        <ResourceDictionary x:Key="{x:Static themes:CommonThemeVariants.LightBlue}">
	    <ColorPaletteResources x:Key="{x:Static themes:CommonThemeVariants.LightBlue}" ... />
        </ResourceDictionary>
        <ResourceDictionary x:Key="{x:Static themes:CommonThemeVariants.DarkBlue}">
            <ColorPaletteResources x:Key="{x:Static themes:CommonThemeVariants.DarkBlue}" ... />
        </ResourceDictionary>
      </ResourceDictionary.ThemeDictionaries>
    </ResourceDictionary>
  </Application.Resources>
```
When I'm setting `Application.Current.RequestedThemeVariant = CommonThemeVariants.LightBlue ` or `Application.Current.RequestedThemeVariant = CommonThemeVariants.DarkBlue` it just switches between original fluent light and dark color palettes.

**Comment by rabbitism** [comment]
Please check this sample. 
https://github.com/AvaloniaUI/Avalonia/blob/master/samples/ControlCatalog/Pages/ThemePage.axaml.cs

**Reply by mysticpillow** [reply]
Thanks for the link. I've learned from these samples how to change specific brushes and colors in runtime by changing `Application.Current.RequestedThemeVariant`. But still it's not clear to me how to switch between multiple custom `ColorPaletteResources` in runtime.

**Comment by maxkatz6** [comment, marked as answer]
FluentTheme has only Dark and Light variants at the moment, and probably ever will. Aside from the possibility to add HighContrast at some point.
FluentTheme.Palletes is a way to override existing palettes for supported variants. It's not a way to add more. I am going to add validation to this dictionary with a clear exception.

Adding more variants is not an intended feature for FluentTheme, but it still should be technically possible without palettes.
As well as custom theme variants are not documented, as they are needed for third-party theme developers and not app developers. You still can create one though.

You need to create a new resource dictionary and include it after FluentTheme. Inside you should have a theme dictionaries for your own variants (referenced with xStatic as you did here).
List of resources you need to put there:
Base colors with adjusted values https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Themes.Fluent/Accents/BaseColorsPalette.xaml#L37-L67
Base brushes duplicated per each variant (as they are statically baked for performance in fluent theme): 
https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Themes.Fluent/Accents/BaseResources.xaml#L218-L377
And control brushes, the same:
https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Themes.Fluent/Accents/FluentControlResources.xaml#L788-L1575


**Keep in mind**. Adding more variants to the FluentTheme is
- not documented
- not well supported
- new brushes/controls might be added mid 11.x and you will need to adjust your values

But still possible, and link from @rabbitism is an smaller example on how to do it https://github.com/AvaloniaUI/Avalonia/blob/master/samples/ControlCatalog/Pages/ThemePage.axaml

**Reply by mysticpillow** [reply]
Thank you for that info. I was able to get this to work :)
